server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractRareTermsAggregator.java

```diff
@@ -0,0 +1,200 @@
+            throw new IllegalStateException("RareTerms agg [" + name() + "] is the child of the nested agg [" + nestedAgg
+                + "], and also has a scoring child agg [" + scoringAgg + "].  This combination is not supported because " +
+                "it requires executing in [depth_first] mode, which the RareTerms agg cannot do.");
+        }
+    }
+
+    @Override
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return true;
+    }
+
+    @Override
+    public DeferringBucketCollector getDeferringCollector() {
+        deferringCollector = new MergingBucketsDeferringCollector(context);
+        return deferringCollector;
+    }
+
+    @Override
+    protected void doPostCollection() {
+        // Make sure we do one final GC to clean up any deleted ords
+        // that may be lingering (but still below GC threshold)
+        gcDeletedEntries(-1);
+    }
+
+    private String subAggsNeedScore() {
+        for (Aggregator subAgg : subAggregators) {
+            if (subAgg.scoreMode().needsScores()) {
+                return subAgg.name();
+            }
+        }
+        return null;
+    }
+
+    private String descendsFromNestedAggregator(Aggregator parent) {
+        while (parent != null) {
+            if (parent.getClass() == NestedAggregator.class) {
+                return parent.name();
+            }
+            parent = parent.parent();
+        }
+        return null;
+    }
+
+    protected void doCollect(V val, int docId) throws IOException {
+        if (filterMightContain(val) == false) {
+            long bucketOrdinal = findOrdinal(val);
+
+            if (bucketOrdinal == -1) {
+                // Brand new term, save into map
+                long ord = addValueToOrds(val);
+                assert ord >= 0;
+                collectBucket(subCollectors, docId, ord);
+
+            } else {
+                // we've seen this value before, see if it is below threshold
+                long termCount = bucketDocCount(bucketOrdinal);
+                if (termCount < maxDocCount) {
+                    // TODO if we only need maxDocCount==1, we could specialize
+                    // and use a bitset instead of a counter scheme
+
+                    collectExistingBucket(subCollectors, docId, bucketOrdinal);
+
+                } else {
+                    // Otherwise we've breached the threshold, add to the cuckoo filter
+                    addValueToFilter(val);
+                    numDeleted += 1;
+
+                    // This is a bit hacky, but we need to collect the value once more to
+                    // make sure the doc_count is over threshold (used later when gc'ing)
+                    collectExistingBucket(subCollectors, docId, bucketOrdinal);
+
+                    if (numDeleted > GC_THRESHOLD) {
+                        gcDeletedEntries(numDeleted);
+                        numDeleted = 0;
+                    }
+                }
+            }
+        }
```

Comment:
Note to self: we may want to check if the value exists in the map here.  E.g. value is in map (sub threshold), but cuckoo throws false positive so we don't check/increment the value in the map which could lead to a "rare" term that shouldn't be.